kotlinx-coroutines-core/common/src/internal/SegmentQueueSynchronizer.kt

```diff
@@ -0,0 +1,606 @@
+ * to simply skip the cell and resume the next waiter, e.g., if this cancelled waiter is the last one.
+ * Thus, it is possible for [SegmentQueueSynchronizer] to refuse this [resume]. In order to support this logic,
+ * users should implement  [onCancellation] function, which returns `true` if the cell can be
+ * moved to the [cancelled][CANCELLED] state, or `false` if the [resume] that comes to this cell
+ * should be refused. In the last case, [tryReturnRefusedValue] is invoked, so that the value
+ * can be put back to the data structure. However, it is possible for [tryReturnRefusedValue] to
+ * fail, and [returnValue] is called in this case. Typically, this [returnValue] function coincide
+ * with the one that resumes waiters (e.g., [release][Semaphore.release] in [Semaphore]). The difference
+ * between [SMART_SYNC] and [SMART_ASYNC] modes is that in the first one the [resume] that comes to the
+ * cell with cancelled waiter waits in a spin-loop until the cell is marked either with [CANCELLED]
+ * or [REFUSE] mark, while in [SMART_ASYNC] mode it replaces the cancelled waiter with the value of
+ * the resumption and finishes, so that the cancellation handler completes this [resume] eventually.
+ * This way, in [SMART_ASYNC] mode the value passed to [resume] can be out of the data structure for a while.
+ *
+ * Here is a state machine for cells. Note that only one [suspend]
+ * and at most one [resume] operation can deal with each cell.
+ *
+ *  +-------+   `suspend` succeeds.   +--------------+  `resume` is   +---------------+  store `RESUMED` to   +---------+  ( `cont` HAS BEEN   )
+ *  |  NULL | ----------------------> | cont: active | -------------> | cont: resumed | --------------------> | RESUMED |  ( RESUMED AND THIS  )
+ *  +-------+                         +--------------+  successful.   +---------------+  avoid memory leaks.  +---------+  ( `resume` SUCCEEDS )
+ *     |                                     |
+ *     |                                     | The continuation
+ *     | `resume` comes to                   | is cancelled.
+ *     | the cell before                     |
+ *     | `suspend` and puts                  V                                                                      ( THE CORRESPONDING `resume` SHOULD BE   )
+ *     | the element into           +-----------------+    The concurrent `resume` should be refused,    +--------+ ( REFUSED AND `tryReturnRefusedValue` OR )
+ *     | the cell, waiting          | cont: cancelled | -----------------------------------------------> | REFUSE | ( `returnValue` IF IT FAILS IS USED TO   )
+ *     | for `suspend` in           +-----------------+      `onCancellation` has returned `false`.      +--------+ ( RETURN THE VALUE BACK TO THE ORIGINAL  )
+ *     | SYNC mode.                          |         \                                                     ^      ( SYNCHRONIZATION PRIMITIVE              )
```

Comment:
> "waiting for `suspend` in SYNC mode

implies that `resume` always waits for `suspend`, while being in some SYNC mode.

", waiting for `suspend` if the resume mode in SYNC", I think, is more clear.